Set undefined as default value for VaParallax target

Sync271 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/epicmaxco/vuestic-ui/blob/master/CODE_OF_CONDUCT.md
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
Fixes #1572 
When it's an empty string, querySelector throws an error as it's only checking if it `string` type or not. Setting the default to `undefined` will let it access the parent element.

## Markup:
<!-- Paste your markup here. -->
```diff
- target = prop<Element | string>({ type: [Object, String], default: '' })
+ target = prop<Element | string | undefined>({ type: [Object, String], default: undefined })
```
<details>

```vue
target = prop<Element | string | undefined>({ type: [Object, String], default: undefined })
```
</details>

in:
https://github.com/epicmaxco/vuestic-ui/blob/8a33fbde75b6092b9842268791267b7b741a1323/packages/ui/src/components/va-parallax/VaParallax.vue#L62-L66
not sure if it has to be changed to the following, when "this.$props.target" is `undefined`, there's no need for `this.$props.target || this.$el.parentElement`:
```vue
  get targetElement () {
    return typeof this.$props.target === 'string'
      ? document.querySelector(this.$props.target)
      : this.$el.parentElement
  }
```

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

m0ksem reviewed (approved):
Thanks!

> not sure if it has to be changed to the following, when "this.$props.target" is undefined, there's no need for this.$props.target || this.$el.parentElement:

User can actually pass HTMLElement instead of string.

---

I'm just curious about this.$el.parentElement... Shouldn't it be a window as default scroll container? 

We can make it even better, if we will try to find the first parent of VaParallax that has overflow scroll property.

Sync271 commented:
> We can make it even better if we will try to find the first parent of VaParallax that has overflow scroll property.

Wouldn't that require the target to accept the `null` type too, since `.parentNode` would return `ParentNode | null` or `Element | null` ?
Wouldn't that end up changing types of a lot of things that depend on the target element?
PS: Sorry, I am new to vue

m0ksem commented:
> > We can make it even better if we will try to find the first parent of VaParallax that has overflow scroll property.
> 
> Wouldn't that require the target to accept the `null` type too, since `.parentNode` would return `ParentNode | null` or `Element | null` ? Wouldn't that end up changing types of a lot of things that depend on the target element? PS: Sorry, I am new to vue

Yes. Also, targetElement can be null on SSR, because nodejs server doesn't have document or window globals, like browser.

Sync271 commented:
> Yes. Also, targetElement can be null on SSR, because nodejs server doesn't have document or window globals, like browser.
Alright, I think I've made the necessary changes, some images for the parallax demo are broken so I used the same image for all cases to test. Let me know if there are any other changes required.

m0ksem commented:
> > Yes. Also, targetElement can be null on SSR, because nodejs server doesn't have document or window globals, like browser.
> 
> Alright, I think I've made the necessary changes, some images for the parallax demo are broken so I used the same image for all cases to test. Let me know if there are any other changes required.

I updated your PR a bit. Removed type casting (`as Element`), added fallback to `document.body` if there is no parent with scroll, added error message if target selector returns null. I also updated our docs demos, because target can be found automatically now 😊. 

Appreciate your help!

LighthouseKeeperYN commented on packages/ui/src/components/va-parallax/VaParallax.vue:
```diff
@@ -59,10 +59,27 @@ export default class VaParallax extends mixins(
+      throw new Error('VaParallax target prop got wrong selector. Target is null')
+    }
+
+    if (this.$props.target) { return this.$props.target }
+
+    return this.getScrollableParent(this.$el.parentElement) || document.body
```
```suggestion
    return this.$props.target || this.getScrollableParent(this.$el.parentElement) || document.body
```

LighthouseKeeperYN commented on packages/ui/src/components/va-parallax/VaParallax.vue:
```diff
@@ -59,10 +59,27 @@ export default class VaParallax extends mixins(
-      : this.$props.target || this.$el.parentElement
+  getScrollableParent (element: Element | null): Element | null {
+    if (element === null) {
+      return null
```
```suggestion
      return window
```

asvae commented:
Thanks.
Doesn't work for me for some reason :<.

<img width="1118" alt="image" src="https://user-images.githubusercontent.com/5394573/159294349-ba78606e-a3f4-49cd-b141-eaa8fb9a1e3e.png">

Sync271 commented:
> Thanks. Doesn't work for me for some reason :<.

Should be fixed now. I implemented the suggested changes without testing, so that's on me. I reverted the changes and fixed the broken `serve:book` links for images.

asvae commented:
@LighthouseKeeperYN can you take a look if we can merge it manually?

LighthouseKeeperYN commented:
Resolved conflicts and merged in #1621

Thanks for your contribution!
